staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go

```diff
@@ -217,26 +217,30 @@ type metricsOptions struct {
 	protocol  string
 }
 
-func (d *dialerCreator) createDialer() utilnet.DialFunc {
+func (d *dialerCreator) createDialerGenerator() func(string) utilnet.DialFunc {
 	if d.direct {
-		return directDialer
-	}
-	return func(ctx context.Context, network, addr string) (net.Conn, error) {
-		trace := utiltrace.New(fmt.Sprintf("Proxy via HTTP Connect over %s", d.options.transport), utiltrace.Field{Key: "address", Value: addr})
-		defer trace.LogIfLong(500 * time.Millisecond)
-		start := egressmetrics.Metrics.Clock().Now()
-		proxier, err := d.connector.connect()
-		if err != nil {
-			egressmetrics.Metrics.ObserveDialFailure(d.options.protocol, d.options.transport, egressmetrics.StageConnect)
-			return nil, err
+		return func(caller string) utilnet.DialFunc {
+			return directDialer
 		}
-		conn, err := proxier.proxy(addr)
-		if err != nil {
-			egressmetrics.Metrics.ObserveDialFailure(d.options.protocol, d.options.transport, egressmetrics.StageProxy)
-			return nil, err
+	}
+	return func(caller string) utilnet.DialFunc {
+		return func(ctx context.Context, network, addr string) (net.Conn, error) {
+			trace := utiltrace.New(fmt.Sprintf("Proxy via HTTP Connect over %s", d.options.transport), utiltrace.Field{Key: "address", Value: addr})
+			defer trace.LogIfLong(500 * time.Millisecond)
+			start := egressmetrics.Metrics.Clock().Now()
+			proxier, err := d.connector.connect(caller)
+			if err != nil {
+				egressmetrics.Metrics.ObserveDialFailure(d.options.protocol, d.options.transport, egressmetrics.StageConnect)
+				return nil, err
+			}
+			conn, err := proxier.proxy(addr, caller)
+			if err != nil {
+				egressmetrics.Metrics.ObserveDialFailure(d.options.protocol, d.options.transport, egressmetrics.StageProxy)
```

Comment:
Would probably be good to use the caller as a label on the metric. We need to be careful that labels on metrics don't get large but that doesn't seem an issue for caller. If we introduced an enum we could put a warning on it.